File: server/routes/settingsRoutes.ts
import { Router, Request, Response } from "express";
import { getStoredGeminiKey, saveGeminiKey, removeGeminiKey, generateWithGeminiAPI } from "../services/geminiService";
import { getStoredCivitaiKey, saveCivitaiKey, removeCivitaiKey } from "../services/civitaiService";
import { getStoredHuggingFaceToken, saveHuggingFaceToken, removeHuggingFaceToken } from "../services/huggingfaceService";
import { getStoredRunpodApiKey, saveRunpodApiKey, removeRunpodApiKey } from "../services/runpodService";
import { getStoredLLMSettings, saveStoredLLMSettings } from "../services/llmSettingsService";
import { detectVisionCapability } from "../services/visionCaptionService";

const router = Router();

/**
 * Get stored program-level LLM settings
 */
router.get("/llm", (req: Request, res: Response) => {
  const settings = getStoredLLMSettings();
  res.json(settings);
});

/**
 * Save program-level LLM settings
 */
router.post("/llm", (req: Request, res: Response) => {
  const settings = saveStoredLLMSettings(req.body || {});
  res.json({ success: true, settings });
});

router.get("/gemini", (req: Request, res: Response) => {
  const key = getStoredGeminiKey();
  const maskedKey = key && key.length > 8 ? `${key.slice(0, 4)}...${key.slice(-4)}` : key ? "***" : null;
  res.json({ configured: !!key, api_key: key ? key.substring(0, 5) + "..." : null, masked_key: maskedKey });
});

router.post("/gemini", (req: Request, res: Response) => {
  const { api_key } = req.body;
  if (!api_key || typeof api_key !== "string" || !api_key.trim()) {
    removeGeminiKey();
  } else {
    saveGeminiKey(api_key.trim());
  }
  res.json({ success: true });
});

router.delete("/gemini", (req: Request, res: Response) => {
  removeGeminiKey();
  res.json({ success: true, message: "Gemini API key removed and deactivated successfully." });
});

/**
 * Get stored Civitai API key status & masked value
 */
router.get("/civitai", (req: Request, res: Response) => {
  const key = getStoredCivitaiKey();
  const maskedKey = key && key.length > 8 ? `${key.slice(0, 4)}...${key.slice(-4)}` : key ? "***" : null;
  res.json({ configured: !!key, api_key: key ? key.substring(0, 5) + "..." : null, masked_key: maskedKey });
});

/**
 * Save Civitai API key
 */
router.post("/civitai", (req: Request, res: Response) => {
  const { api_key, apiKey } = req.body || {};
  const val = (api_key || apiKey || "").trim();
  if (val) {
    saveCivitaiKey(val);
  } else {
    removeCivitaiKey();
  }
  res.json({ success: true });
});

/**
 * Delete Civitai API key
 */
router.delete("/civitai", (req: Request, res: Response) => {
  removeCivitaiKey();
  res.json({ success: true, message: "Civitai API key removed." });
});

/**
 * Get stored Hugging Face token status & masked value
 */
router.get("/huggingface", (req: Request, res: Response) => {
  const token = getStoredHuggingFaceToken();
  const maskedToken = token && token.length > 8 ? `${token.slice(0, 4)}...${token.slice(-4)}` : token ? "***" : null;
  res.json({ configured: !!token, token: token ? token.substring(0, 5) + "..." : null, masked_token: maskedToken });
});

/**
 * Save Hugging Face token
 */
router.post("/huggingface", (req: Request, res: Response) => {
  const { token, api_token, apiKey } = req.body || {};
  const val = (token || api_token || apiKey || "").trim();
  if (val) {
    saveHuggingFaceToken(val);
  } else {
    removeHuggingFaceToken();
  }
  res.json({ success: true });
});

/**
 * Delete Hugging Face token
 */
router.delete("/huggingface", (req: Request, res: Response) => {
  removeHuggingFaceToken();
  res.json({ success: true, message: "Hugging Face token removed." });
});

/**
 * Get stored RunPod API key status & masked value
 */
router.get("/runpod", (req: Request, res: Response) => {
  const key = getStoredRunpodApiKey();
  const maskedKey = key && key.length > 8 ? `${key.slice(0, 4)}...${key.slice(-4)}` : key ? "***" : null;
  res.json({ configured: !!key, api_key: key || null, masked_key: maskedKey });
});

/**
 * Save RunPod API key
 */
router.post("/runpod", (req: Request, res: Response) => {
  const { runpod_api_key, api_key, apiKey } = req.body || {};
  const val = (runpod_api_key || api_key || apiKey || "").trim();
  if (val) {
    saveRunpodApiKey(val);
  } else {
    removeRunpodApiKey();
  }
  res.json({ success: true });
});

/**
 * Delete RunPod API key
 */
router.delete("/runpod", (req: Request, res: Response) => {
  removeRunpodApiKey();
  res.json({ success: true, message: "RunPod API key removed." });
});

/**
 * Test Local LLM (LM Studio / Ollama / OpenAI-compatible) API endpoint connection
 */
router.post(["/test-lm-studio", "/test-local-llm"], async (req: Request, res: Response) => {
  const { url, endpoint, targetUrl: rawTargetUrl } = req.body || {};
  const inputUrl = url || endpoint || rawTargetUrl || "http://localhost:1234/v1";
  const targetUrl = inputUrl.trim().replace(/\/$/, "");

  let probeUrl = targetUrl;
  if (!probeUrl.endsWith("/models")) {
    if (probeUrl.endsWith("/v1")) {
      probeUrl = `${probeUrl}/models`;
    } else {
      probeUrl = `${probeUrl}/v1/models`;
    }
  }

  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 5000);

    const lmRes = await fetch(probeUrl, {
      method: "GET",
      headers: { "Accept": "application/json" },
      signal: controller.signal
    });
    clearTimeout(timeout);

    if (lmRes.ok) {
      const data = await lmRes.json().catch(() => ({}));
      const modelsList = Array.isArray(data.data) ? data.data : (Array.isArray(data.models) ? data.models : []);
      const modelNames = modelsList
        .map((m: any) => m.id || m.name || (typeof m === "string" ? m : ""))
        .filter(Boolean);
      const modelsCount = modelNames.length;

      // Auto-detect vision support across available models
      const { hasVision, visionModel } = detectVisionCapability(modelsList.length > 0 ? modelsList : modelNames);

      // Auto-detect backend engine
      const serverHeader = (lmRes.headers.get("server") || "").toLowerCase();
      const isOllama =
        targetUrl.includes("11434") ||
        serverHeader.includes("ollama") ||
        modelNames.some((n: string) => n.includes(":") && !n.includes("@"));

      const isLmStudio = targetUrl.includes("1234") || serverHeader.includes("lmstudio");
      const backendType: "ollama" | "lm_studio" | "generic" = isOllama
        ? "ollama"
        : isLmStudio
        ? "lm_studio"
        : "generic";

      const backendDisplayName = isOllama ? "Ollama" : isLmStudio ? "LM Studio" : "Local LLM";

      return res.json({
        success: true,
        backend: backendType,
        message: isOllama
          ? `Ollama connected (${modelsCount} model${modelsCount === 1 ? "" : "s"} available)`
          : `${backendDisplayName} server responsive at ${targetUrl}`,
        modelsCount,
        models: modelNames,
        modelNames: modelNames.slice(0, 5).join(", "),
        hasVision,
        visionModel,
        probeUrl
      });
    } else {
      return res.status(400).json({
        success: false,
        error: `Server responded with HTTP ${lmRes.status} ${lmRes.statusText}`
      });
    }
  } catch (err: any) {
    const isAbort = err.name === "AbortError";
    const errorMessage = isAbort 
      ? "Connection timed out (5s limit reached)" 
      : (err.message || "Connection refused or endpoint unreachable");

    return res.status(400).json({
      success: false,
      error: errorMessage
    });
  }
});

/**
 * Test Google Gemini API connection with lightweight verification query
 */
router.post("/test-gemini", async (req: Request, res: Response) => {
  let { api_key, apiKey } = req.body || {};
  let keyToUse = (api_key || apiKey || "").trim();

  if (!keyToUse) {
    keyToUse = getStoredGeminiKey() || "";
  }

  if (!keyToUse) {
    return res.status(400).json({
      success: false,
      error: "No Gemini API key is configured"
    });
  }

  try {
    const result = await generateWithGeminiAPI(keyToUse, "Ping test connection verification");
    if (result && result.text) {
      return res.json({
        success: true,
        message: `Gemini API key verified successfully using ${result.modelUsed}`,
        activeModel: result.modelUsed,
        modelUsed: result.modelUsed
      });
    } else {
      return res.status(400).json({
        success: false,
        error: "Gemini API returned an empty response"
      });
    }
  } catch (err: any) {
    let errMsg = err.message || "Invalid API key or network request failed";
    if (errMsg.toLowerCase().includes("api_key_invalid") || errMsg.toLowerCase().includes("api key not valid") || errMsg.toLowerCase().includes("unauthorized") || errMsg.toLowerCase().includes("403") || errMsg.toLowerCase().includes("401")) {
      errMsg = "Invalid API key provided";
    }
    return res.status(400).json({
      success: false,
      error: errMsg
    });
  }
});

/**
 * Test ComfyUI endpoint connection
 */
router.post("/test-comfyui", async (req: Request, res: Response) => {
  const { url, comfyui_url, comfyui_api_url, token, remote_api_token } = req.body;
  const rawUrl = comfyui_url || url || comfyui_api_url || "http://127.0.0.1:8188";
  const targetUrl = rawUrl.trim().replace(/\/$/, "");
  const authToken = (remote_api_token || token || "").trim();

  console.log(`[ComfyUI Test] Testing reachability for URL: ${targetUrl}`);

  let probeUrl = `${targetUrl}/system_stats`;

  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 5000);

    const headers: Record<string, string> = { "Accept": "application/json" };
    if (authToken) {
      headers["Authorization"] = `Bearer ${authToken}`;
    }

    let lmRes = await fetch(probeUrl, {
      method: "GET",
      headers,
      signal: controller.signal
    }).catch(async () => {
      // Fallback probe
      probeUrl = `${targetUrl}/object_info`;
      return await fetch(probeUrl, {
        method: "GET",
        headers,
        signal: controller.signal
      });
    });
    clearTimeout(timeout);

    if (lmRes && lmRes.ok) {
      const data = await lmRes.json().catch(() => ({}));
      
      const systemInfoParts: string[] = [];
      if (data && data.system && data.system.os) {
        systemInfoParts.push(`OS: ${data.system.os}`);
      }
      if (data && data.devices && data.devices.length > 0) {
        const device = data.devices[0];
        let devStr = device.name ? `GPU: ${device.name}` : "GPU detected";
        if (device.vram_total) {
          const vramGB = (device.vram_total / (1024 * 1024 * 1024)).toFixed(1);
          devStr += ` (${vramGB}GB VRAM)`;
        }
        systemInfoParts.push(devStr);
      }

      const systemInfo = systemInfoParts.length > 0 ? systemInfoParts.join(" | ") : `${targetUrl} (Active)`;
      console.log(`[ComfyUI Test] Success (HTTP 200) - Connected to: ${systemInfo}`);

      return res.json({
        success: true,
        message: `ComfyUI server responsive at ${targetUrl}`,
        systemInfo,
        probeUrl,
        system_stats: data
      });
    } else {
      const statusText = lmRes ? `HTTP ${lmRes.status} ${lmRes.statusText}` : "No response";
      const errorMessage = `Server responded with ${statusText}`;
      console.log(`[ComfyUI Test] Failed to connect to ${targetUrl}: ${errorMessage}`);
      return res.status(400).json({
        success: false,
        error: errorMessage
      });
    }
  } catch (err: any) {
    const isAbort = err.name === "AbortError";
    const errorMessage = isAbort 
      ? "Connection timed out (5s limit reached)" 
      : (err.message || "Connection refused or endpoint unreachable");

    console.log(`[ComfyUI Test] Failed to connect to ${targetUrl}: ${errorMessage}`);
    return res.status(400).json({
      success: false,
      error: errorMessage
    });
  }
});

export default router;

File: backend/routes/settings_routes.py
import os
import json
import httpx
from typing import Dict, Any, List, Optional
from fastapi import APIRouter, HTTPException
from fastapi.responses import JSONResponse
from pydantic import BaseModel

from backend.utils.file_handlers import ASSETS_DIR
from backend.services.civitai_service import get_stored_civitai_key, save_civitai_key
from backend.services.huggingface_service import get_stored_huggingface_token, save_huggingface_token

router = APIRouter(prefix="/settings", tags=["Application Settings"])

class CivitaiSettingsRequest(BaseModel):
    api_key: Optional[str] = None
    apiKey: Optional[str] = None

    class Config:
        extra = "allow"

class HuggingFaceSettingsRequest(BaseModel):
    token: Optional[str] = None
    huggingface_token: Optional[str] = None
    api_token: Optional[str] = None
    apiKey: Optional[str] = None

    class Config:
        extra = "allow"

class GeminiSettingsRequest(BaseModel):
    api_key: Optional[str] = None
    apiKey: Optional[str] = None

    class Config:
        extra = "allow"

class LMStudioTestRequest(BaseModel):
    url: Optional[str] = None
    lm_studio_url: Optional[str] = None
    endpoint: Optional[str] = None
    targetUrl: Optional[str] = None

    class Config:
        extra = "allow"

class GeminiTestRequest(BaseModel):
    api_key: Optional[str] = None
    apiKey: Optional[str] = None

    class Config:
        extra = "allow"

class ComfyUITestRequest(BaseModel):
    url: Optional[str] = None
    comfyui_url: Optional[str] = None
    comfyui_api_url: Optional[str] = None
    token: Optional[str] = None
    remote_api_token: Optional[str] = None

    class Config:
        extra = "allow"


@router.get("/civitai")
async def get_civitai_settings():
    """Read stored Civitai API key and return masked status."""
    key = get_stored_civitai_key()
    masked_key = f"{key[:4]}...{key[-4:]}" if len(key) > 8 else ("***" if key else None)
    return {
        "configured": bool(key),
        "api_key": f"{key[:5]}..." if key else None,
        "masked_key": masked_key
    }

@router.post("/civitai")
async def save_civitai_settings(req: CivitaiSettingsRequest):
    """Save Civitai API key to assets/civitai_config.json."""
    key = (req.api_key or req.apiKey or "").strip()
    try:
        save_civitai_key(key)
        return {"success": True}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/huggingface")
async def get_huggingface_settings():
    """Read stored Hugging Face token and return masked status."""
    token = get_stored_huggingface_token()
    masked_token = f"{token[:4]}...{token[-4:]}" if len(token) > 8 else ("***" if token else None)
    return {
        "configured": bool(token),
        "token": f"{token[:5]}..." if token else None,
        "masked_token": masked_token
    }

@router.post("/huggingface")
async def save_huggingface_settings(req: HuggingFaceSettingsRequest):
    """Save Hugging Face token to assets/huggingface_config.json."""
    token = (req.token or req.huggingface_token or req.api_token or req.apiKey or "").strip()
    try:
        save_huggingface_token(token)
        return {"success": True}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.get("/gemini")
async def get_gemini_settings():
    """Retrieve current Gemini API configuration state."""
    gemini_file = ASSETS_DIR / "gemini_config.json"
    key = None
    if gemini_file.exists():
        try:
            with open(gemini_file, "r", encoding="utf-8") as f:
                data = json.load(f)
                key = data.get("api_key")
        except Exception:
            pass
    if not key:
        key = os.environ.get("GEMINI_API_KEY")
    masked_key = f"{key[:4]}...{key[-4:]}" if key and len(key) > 8 else ("***" if key else None)
    return {
        "configured": bool(key),
        "api_key": f"{key[:5]}..." if key else None,
        "masked_key": masked_key
    }

@router.post("/gemini")
async def save_gemini_settings(req: GeminiSettingsRequest):
    """Save Gemini API key to local configuration."""
    key = (req.api_key or req.apiKey or "").strip()
    gemini_file = ASSETS_DIR / "gemini_config.json"
    try:
        ASSETS_DIR.mkdir(parents=True, exist_ok=True)
        with open(gemini_file, "w", encoding="utf-8") as f:
            json.dump({"api_key": key}, f, indent=2)
        return {"success": True}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

@router.post("/test-lm-studio")
async def test_lm_studio(req: LMStudioTestRequest):
    """Test LM Studio endpoint reachability and list available models."""
    raw_url = req.url or req.lm_studio_url or req.endpoint or req.targetUrl or "http://localhost:1234/v1"
    url = raw_url.strip().rstrip("/")

    if not url.endswith("/models"):
        if url.endswith("/v1"):
            probe_url = f"{url}/models"
        else:
            probe_url = f"{url}/v1/models"
    else:
        probe_url = url

    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            res = await client.get(probe_url)
            if res.status_code == 200:
                data = res.json()
                models_data = data.get("data", []) if isinstance(data, dict) else []
                count = len(models_data)
                names_list = [m.get("id") or m.get("name") or str(m) for m in models_data if isinstance(m, dict)]
                model_names_str = ", ".join(names_list[:3]) if names_list else "None"
                return {
                    "success": True,
                    "message": f"LM Studio server responsive at {url}",
                    "modelsCount": count,
                    "models": names_list,
                    "modelNames": model_names_str,
                    "probeUrl": probe_url
                }
            else:
                err_msg = f"HTTP {res.status_code} - {res.text[:200]}"
                return JSONResponse(status_code=400, content={"success": False, "error": f"Failed to connect to LM Studio: {err_msg}"})
    except Exception as e:
        return JSONResponse(status_code=400, content={"success": False, "error": f"Failed to connect to {url}: {str(e)}"})

@router.post("/test-gemini")
async def test_gemini(req: GeminiTestRequest):
    """Test Gemini API key validity with a lightweight test query."""
    key_to_use = (req.api_key or req.apiKey or "").strip()

    if not key_to_use:
        gemini_file = ASSETS_DIR / "gemini_config.json"
        if gemini_file.exists():
            try:
                with open(gemini_file, "r", encoding="utf-8") as f:
                    data = json.load(f)
                    key_to_use = (data.get("api_key") or "").strip()
            except Exception:
                pass
        if not key_to_use:
            key_to_use = (os.environ.get("GEMINI_API_KEY") or "").strip()

    if not key_to_use:
        return JSONResponse(status_code=400, content={"success": False, "error": "No Gemini API key is configured"})

    model_name = "gemini-3.7-flash"
    endpoint = f"https://generativelanguage.googleapis.com/v1beta/models/{model_name}:generateContent?key={key_to_use}"
    payload = {"contents": [{"parts": [{"text": "Ping test connection verification"}]}]}

    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            res = await client.post(endpoint, json=payload)
            if res.status_code == 200:
                return {
                    "success": True,
                    "message": f"Gemini API key verified successfully using {model_name}",
                    "activeModel": model_name,
                    "modelUsed": model_name
                }
            else:
                err_data = res.json() if res.headers.get("content-type", "").startswith("application/json") else {}
                err_msg = err_data.get("error", {}).get("message") or f"HTTP {res.status_code}: {res.text[:200]}"
                return JSONResponse(status_code=400, content={"success": False, "error": err_msg})
    except Exception as e:
        return JSONResponse(status_code=400, content={"success": False, "error": str(e)})

@router.post("/test-comfyui")
async def test_comfyui(req: ComfyUITestRequest):
    """Test ComfyUI endpoint reachability and extract system/device stats."""
    raw_url = req.comfyui_url or req.url or req.comfyui_api_url or "http://127.0.0.1:8188"
    url = raw_url.strip().rstrip("/")
    token = (req.remote_api_token or req.token or "").strip()

    headers = {"Accept": "application/json"}
    if token:
        headers["Authorization"] = f"Bearer {token}"

    probe_url = f"{url}/system_stats"
    try:
        async with httpx.AsyncClient(timeout=5.0) as client:
            try:
                res = await client.get(probe_url, headers=headers)
            except Exception:
                probe_url = f"{url}/object_info"
                res = await client.get(probe_url, headers=headers)

            if res.status_code == 200:
                data = {}
                try:
                    data = res.json()
                except Exception:
                    pass

                system_info_parts = []
                if isinstance(data, dict):
                    sys_info = data.get("system", {})
                    if isinstance(sys_info, dict) and sys_info.get("os"):
                        system_info_parts.append(f"OS: {sys_info['os']}")

                    devices = data.get("devices", [])
                    if isinstance(devices, list) and len(devices) > 0:
                        first_dev = devices[0]
                        if isinstance(first_dev, dict):
                            dev_name = first_dev.get("name")
                            dev_str = f"GPU: {dev_name}" if dev_name else "GPU detected"
                            vram_total = first_dev.get("vram_total")
                            if vram_total and isinstance(vram_total, (int, float)):
                                vram_gb = f"{vram_total / (1024 * 1024 * 1024):.1f}GB VRAM"
                                dev_str += f" ({vram_gb})"
                            system_info_parts.append(dev_str)

                device_info = " | ".join(system_info_parts) if system_info_parts else f"{url} (Active)"
                return {
                    "success": True,
                    "message": f"ComfyUI server responsive at {url}",
                    "systemInfo": device_info,
                    "probeUrl": probe_url,
                    "system_stats": data
                }
            else:
                err_msg = f"Server responded with HTTP {res.status_code}"
                return JSONResponse(status_code=400, content={"success": False, "error": err_msg})
    except Exception as e:
        return JSONResponse(status_code=400, content={"success": False, "error": str(e)})
